h5py/tests/hl/test_datatype.py

```diff
@@ -340,6 +340,37 @@ def test_fixed_ascii(self):
         assert string_info.length == 10
         assert h5py.check_vlen_dtype(dt) is None
 
+class TestDateTime(TestCase):
+    def test_datetime(self):
+        dt_units = [
+            # Dates
+            '[Y]', '[M]', '[D]',
+            # Times
+            '[h]', '[m]', '[s]', '[ms]', '[us]',
+            '[ns]', '[ps]', '[fs]', '[as]',
+        ]
+
+        for dt_kind in ['M8', 'm8']:
+            for dt_unit in dt_units:
+                for dt_order in ['<', '>']:
+                    dt_descr = dt_order + dt_kind + dt_unit
+                    dt = np.dtype(dt_descr)
+                    h5py.register_dtype(dt)
+                    self.assertRaises(RuntimeError, h5py.register_dtype, dt)
+                    fname = self.mktemp()
+
+                    arr = (np.array([0], dtype=np.int64).view(dtype=dt)
+                        if dt_kind == 'M8'
+                        else np.array([np.timedelta64(500, dt_unit[1:-1])], dtype=dt))
+
+                    with h5py.File(fname, 'w') as f:
+                        dset = f.create_dataset("default", data=arr)
+                        self.assertArrayEqual(arr, dset)
+                        self.assertEqual(arr.dtype, dset.dtype)
+                        # So both branches are tested.
+                        h5py.deregister_dtype(dt)
```

Comment:
At a minimum, we should put the deregister in a `finally` block so it still happens if something goes wrong during the test.

But this also highlights that global mutable state is always a tricky thing. I'm still inclined to postpone this for a future release so we can think about how best to design it rather than rushing to get it in before 2.10.

Reply:
I tend to agree here... My concern is Py2 compat. I can use thread_local if needed.

Reply:
I put it in `finally`.

Reply:
I really would like for this to go into 2.10.  If we think more effort and discussion is needed to be done to make this acceptable for 2.10, @hameerabbasi and I are happy to go through that effort.

I agree that a context manager would be nice for this, so we wouldn't have to worry about using try-finallys everywhere.

Maybe we could use the thread_local behavior on Python 3 and then not use that for Python 2, with a warning in the docs that the context manager might behave slightly differently on Python 2 and explain why.  Presumably, we aren't planning on supporting Python 2 forever, so this issue will go away in the relatively near future.

Reply:
The plan is to drop Python 2 for 3.0 - see issue #1177. That will probably be the next release after 2.10.

Rather than some (thread-)global registry, would it make sense to require manually creating wrapper objects in a similar pattern to strings, vlens and enums? Something like:

```python
dt = h5py.opaque_dtype(arr.dtype)
dset = f.create_dataset('timestamps', data=arr, dtype=dt)
```

Reply:
Would we be able to put such a type inside a compound type?

Reply:
I imagine it would be possible, but perhaps it's a bit ugly if you have to define the compound structure again for h5py. On the other hand, it would make it explicit how the opaque data type is being used.